lib/ClangImporter/ClangModuleDependencyScanner.cpp

```diff
@@ -0,0 +1,341 @@
+      continue;
+
+    // File dependencies for this module.
+    std::vector<std::string> fileDeps;
+    for (const auto &fileDep : clangModuleDep.FileDeps) {
+      fileDeps.push_back(fileDep.getKey());
+    }
+
+    // Module-level dependencies.
+    llvm::StringSet<> alreadyAddedModules;
+    auto dependencies = ModuleDependencies::forClangModule(
+        clangModuleDep.ImplicitModulePCMPath,
+        clangModuleDep.ClangModuleMapFile,
+        clangModuleDep.ContextHash,
+        clangModuleDep.NonPathCommandLine,
+        fileDeps);
+    for (const auto &moduleName : clangModuleDep.ClangModuleDeps) {
+      dependencies.addModuleDependency(moduleName.ModuleName, alreadyAddedModules);
+    }
+
+    cache.recordDependencies(clangModuleDep.ModuleName,
+                             std::move(dependencies),
+                             ModuleDependenciesKind::Clang);
+  }
+}
+
+Optional<ModuleDependencies> ClangImporter::getModuleDependencies(
+    StringRef moduleName, ModuleDependenciesCache &cache) {
+  // Check whether there is already a cached result.
+  if (auto found = cache.findDependencies(
+          moduleName, ModuleDependenciesKind::Clang))
+    return found;
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
+
+  // HACK! Replace the module import buffer name with the source file hack.
+  auto importHackFile = clangImpl->getImportHackFile();
+  if (!importHackFile) {
+    // FIXME: Emit a diagnostic here.
+    return None;
+  }
+
+  // Reform the Clang importer options.
+  // FIXME: Just save a reference or copy so we can get this back.
+  ClangImporterOptions importerOpts;
+
+  // Determine the command-line arguments for dependency scanning.
+  auto &ctx = Impl.SwiftContext;
+  std::vector<std::string> commandLineArgs =
+    getClangDepScanningInvocationArguments(
+      ctx, importerOpts, *importHackFile);
+
+  // HACK! Trick out a .m file to use to import the module we name.
+  std::string moduleNameHackDefine =
+      ("-DHACK_MODULE_NAME=" + moduleName).str();
+  commandLineArgs.push_back(moduleNameHackDefine);
+  commandLineArgs.push_back("-fmodules-ignore-macro=HACK_MODULE_NAME");
+
+  std::string workingDir =
+      ctx.SourceMgr.getFileSystem()->getCurrentWorkingDirectory().get();
+  CompileCommand command(workingDir, *importHackFile, commandLineArgs, "-");
+  SingleCommandCompilationDatabase database(command);
+
+  auto clangDependencies = clangImpl->tool.getFullDependencies(
+      database, workingDir, clangImpl->alreadySeen);
+
+  if (!clangDependencies) {
+    // FIXME: Route this to a normal diagnostic.
+    llvm::logAllUnhandledErrors(clangDependencies.takeError(), llvm::errs());
+    return None;
+  }
+
+  // Record module dependencies for each module we found.
+  recordModuleDependencies(cache, *clangDependencies);
+
+  return cache.findDependencies(moduleName, ModuleDependenciesKind::Clang);
+}
+
+bool ClangImporter::addBridgingHeaderDependencies(
```

Comment:
You know me; I would prefer to break up this function into more understandable (for me) pieces.